HUDI-138 - Meta Files handling also need to support consistency guard

bvaradar opened the pull request.

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/HoodieTableMetaClient.java:
```diff
@@ -79,13 +86,23 @@ public HoodieTableMetaClient(Configuration conf, String basePath)
+    this(conf, basePath, loadActiveTimelineOnLoad, false, 0, 0, 0);
+  }
+
+  public HoodieTableMetaClient(Configuration conf, String basePath,
```
This is polluting the constructor, can we wrap all consistency check configs into a single config and wire them here ?

bvaradar replied:
Fixed

bvaradar commented on NOTICE.txt:
```diff
@@ -8,14 +8,13 @@ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 
 This project includes:
+  aircompressor under Apache License 2.0
```
I think these changes might be because of non-strict package dependencies as there are no POM related changes here.

bvaradar replied:
Removed

bvaradar commented:
@n3nash : Addressed review comments

vinothchandar commented on hoodie-cli/src/main/java/com/uber/hoodie/cli/HoodieCLI.java:
```diff
@@ -37,6 +40,15 @@
+
+  public static void setBasePath(String basePath) {
+    HoodieCLI.basePath = basePath;
+    new HoodieTableMetaClient(HoodieCLI.conf, basePath, false, HoodieCLI.consistencyGuardConfig);
```
are we throwing away this object?

vinothchandar commented on hoodie-cli/src/main/java/com/uber/hoodie/cli/HoodieCLI.java:
```diff
@@ -47,11 +59,11 @@ public static boolean initConf() {
-  public static void setTableMetadata(HoodieTableMetaClient tableMetadata) {
-    HoodieCLI.tableMetadata = tableMetadata;
+  public static void refreshTableMetadata() {
+    setBasePath(basePath);
```
this might be a better place to reinit tableMetaClient than setBasePath() ?

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/io/storage/HoodieWrapperFileSystem.java:
```diff
@@ -236,7 +237,28 @@ public FSDataOutputStream append(Path f, int bufferSize, Progressable progress)
+      try {
+        consistencyGuard.waitTillFileDisappears(convertToDefaultPath(src));
+      } catch (TimeoutException e) {
+        throw new HoodieException("Timed out waiting for " + dst + " to disappear", e);
```
src to disappear?

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/HoodieTableMetaClient.java:
```diff
@@ -190,13 +201,27 @@ public HoodieTableConfig getTableConfig() {
     if (fs == null) {
-      fs = FSUtils.getFs(metaPath, hadoopConf.get());
+      FileSystem fileSystem = FSUtils.getFs(metaPath, hadoopConf.get());
+      if (fileSystem instanceof HoodieWrapperFileSystem) {
```
when you anticipate fs to be the wrapper filesystem ?

bvaradar replied:
This was a defensive check as the effect of nesting could result in longer wait times. With  the current code, this should not happen. I can replace with a precondition that we should never see a HoodieWrapperFileSystem. Let me know how it sounds ?

vinothchandar replied:
yes. We can do a PreCondition if this should not happen.. Rather hit the issue, debug and fix it, than living dangerously :)

bvaradar replied:
Added precondition.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/HoodieTableMetaClient.java:
```diff
@@ -223,6 +248,13 @@ public synchronized HoodieActiveTimeline reloadActiveTimeline() {
   }
 
+  public ConsistencyGuardConfig getConsistencyGuardConfig() {
+    if (null == consistencyGuardConfig) {
```
L80 above should ensure this is non-null?

bvaradar replied:
Fixed.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/io/storage/HoodieWrapperFileSystem.java:
```diff
@@ -377,7 +399,15 @@ public FSDataOutputStream createNonRecursive(Path f, FsPermission permission,
   @Override
   public boolean createNewFile(Path f) throws IOException {
-    return fileSystem.createNewFile(convertToDefaultPath(f));
+    boolean newFile = fileSystem.createNewFile(convertToDefaultPath(f));
```
should we cover `mkdirs` as well? also createNonRecursive and other write methods? Can you skim this file for any write operation that we may have to consider?

bvaradar replied:
Fixed

vinothchandar commented:
@bvaradar is this ready for merging?

bvaradar commented:
@n3nash @vinothchandar : Have addressed all the comments and ran automated demo on top of this change. As discussed earlier, I am going ahead with merging this change.
